gremlin-dotnet/src/Gremlin.Net/Driver/ConnectionPool.cs

```diff
@@ -33,91 +34,134 @@ internal class ConnectionPool : IDisposable
+            }
+
+            Interlocked.CompareExchange(ref _nrConnections, _minPoolSize, 0);
+        }
 
         public async Task<IConnection> GetAvailableConnectionAsync()
         {
-            if (!TryGetConnectionFromPool(out var connection))
-                connection = await CreateNewConnectionAsync().ConfigureAwait(false);
+            if (TryGetConnectionFromPool(out var connection))
+                return ProxiedConnection(connection);
+            connection = await AddConnectionIfUnderMaximumAsync().ConfigureAwait(false) ?? WaitForConnection();
+            return ProxiedConnection(connection);
+        }
 
-            return new ProxyConnection(connection, AddConnectionIfOpen);
+        private IConnection ProxiedConnection(Connection connection)
+        {
+            return new ProxyConnection(connection, ReturnConnectionIfOpen);
         }
 
-        private bool TryGetConnectionFromPool(out Connection connection)
+        private void ReturnConnectionIfOpen(Connection connection)
+        {
+            if (!connection.IsOpen)
+            {
+                ConsiderUnavailable();
+                DefinitelyDestroyConnection(connection);
+                return;
+            }
+
+            _connections.Add(connection);
+            _newConnectionAvailable.Set();
+        }
+
+        private async Task<Connection> AddConnectionIfUnderMaximumAsync()
         {
             while (true)
             {
-                connection = null;
-                lock (_connectionsLock)
-                {
-                    if (_connections.IsEmpty) return false;
-                    _connections.TryTake(out connection);
-                }
+                var nrOpened = Interlocked.CompareExchange(ref _nrConnections, 0, 0);
+                if (nrOpened >= _maxPoolSize) return null;
 
-                if (connection.IsOpen) return true;
-                connection.Dispose();
+                if (Interlocked.CompareExchange(ref _nrConnections, nrOpened + 1, nrOpened) == nrOpened)
+                    break;
             }
+
+            return await CreateNewConnectionAsync().ConfigureAwait(false);
         }
 
         private async Task<Connection> CreateNewConnectionAsync()
         {
-            NrConnections++;
             var newConnection = _connectionFactory.CreateConnection();
             await newConnection.ConnectAsync().ConfigureAwait(false);
             return newConnection;
         }
 
-        private void AddConnectionIfOpen(Connection connection)
+        private Connection WaitForConnection()
         {
-            if (!connection.IsOpen)
+            var start = DateTimeOffset.Now;
+            var remaining = _waitForConnectionTimeout;
+            do
             {
-                ConsiderUnavailable();
-                connection.Dispose();
-                return;
-            }
-            AddConnection(connection);
+                if (_newConnectionAvailable.WaitOne(remaining))
```

Comment:
This loop blocks the calling thread, which is a problem as we would be mixing async and blocking code.

We should implemented in a way that it awaits for a connection to be available. Here are some good starting points (I'm not aware of a built-in mechanism in .NET): 
- https://github.com/StephenCleary/AsyncEx/wiki/AsyncAutoResetEvent
- https://blogs.msdn.microsoft.com/pfxteam/2012/02/11/building-async-coordination-primitives-part-2-asyncautoresetevent/

Reply:
Thanks for pointing this out @jorgebay 

I actually experimented with something like an `AutoResetEvent` for the very first version of Gremlin.Net, but didn't get it to work and therefore ended up with the plain `lock`. I guess what I was missing back then was simply the knowledge about the [CAS operation](https://en.wikipedia.org/wiki/Compare-and-swap) and how it can be used to avoid synchronous locks.
I'll give this a try again to come up with a version that hopefully completely removes synchronous blocking. I guess this will result in our own version of an `AsyncAutoResetEvent` that allows us to wait with a timeout as that seems to be missing in both existing implementations (from the blog and the library you linked). (I'm also not sure whether it's worth it to add a dependency on `AsyncEx` just to get one class.)

Reply:
I would prefer getting the `AsyncAutoResetEvent` source code (If the license permits it) or develop our own version, than adding a dependency.

Reply:
I finally found some time to work on this again and implemented an `AsyncAutoResetEvent` that is based on the blog post by Stephen Toub. The main difference to the version in the blog post is that it only accepts waits that specify a timeout and it uses a `List` as the underlying data structure instead of a `Queue` as we need to remove wait tasks from it that timed out.
Would be good if you could give this a thorough review as it's probably easy to make small mistakes in a primitive like this that will be hard to find later on.